octodns/source/axfr.py

```diff
@@ -0,0 +1,157 @@
+
+    def _data_for_single(self, _type, records):
+        record = records[0]
+        return {
+            'ttl': record['ttl'],
+            'type': _type,
+            'value': record['value']
+        }
+
+    _data_for_CNAME = _data_for_single
+    _data_for_PTR = _data_for_single
+
+    def _data_for_SRV(self, _type, records):
+        values = []
+        for record in records:
+            priority, weight, port, target = record['value'].split(' ', 3)
+            values.append({
+                'priority': priority,
+                'weight': weight,
+                'port': port,
+                'target': target,
+            })
+        return {
+            'type': _type,
+            'ttl': records[0]['ttl'],
+            'values': values
+        }
+
+    def populate(self, zone, target=False, lenient=False):
+        self.log.debug('populate: name=%s, target=%s, lenient=%s', zone.name,
+                       target, lenient)
+
+        values = defaultdict(lambda: defaultdict(list))
+        for record in self.zone_records(zone):
+            name = zone.hostname_from_fqdn(record['name'])
+            _type = record['type']
+            values[name][record['type']].append(record)
+
+        before = len(zone.records)
+        for name, types in values.items():
+            for _type, records in types.items():
+                data_for = getattr(self, '_data_for_{}'.format(_type))
+                record = Record.new(zone, name, data_for(_type, records),
+                                    source=self, lenient=lenient)
+                zone.add_record(record, lenient=lenient)
+
+        self.log.info('populate:   found %s records',
+                      len(zone.records) - before)
+
+
+class AxfrSourceException(Exception):
+    pass
+
+
+class AxfrSourceZoneTransferFailed(AxfrSourceException):
+
+    def __init__(self):
+        super(AxfrSourceZoneTransferFailed, self).__init__('AXFR failed')
+
+
+class AxfrSource(AxfrBaseSource):
+    '''
+    Axfr zonefile importer to import data
+
+    axfr:
+        class: octodns.source.axfr.AxfrSource
+        # The address of nameserver to perform zone transfer against
+        master: ns1.example.com
+    '''
+    def __init__(self, id, master):
+        self.log = logging.getLogger('AxfrSource[{}]'.format(id))
+        self.log.debug('__init__: id=%s, master=%s', id, master)
+        super(AxfrSource, self).__init__(id)
+        self.master = master
+
+    def zone_records(self, zone):
+        try:
+            zone = dns.zone.from_xfr(dns.query.xfr(self.master, zone.name,
+                                                   relativize=False),
+                                     relativize=False)
+        except DNSException:
+            raise AxfrSourceZoneTransferFailed()
+
+        records = []
+
+        for (name, ttl, rdata) in zone.iterate_rdatas():
+            rdtype = dns.rdatatype.to_text(rdata.rdtype)
+            if rdtype != 'SOA':
```

Comment:
Should this use `SUPPORTS` to filter things to only the supported record types?

Reply:
I was thinking about this today, I think I would end up removing it here, and adding it within the populate method.

Ideally this `zone_records` method is just the thing that performs the zone transfer and transforms the response into a more sane data structure, it shouldn't care about filtering records by type.

However `populate` should care about what record types get let through.

I will make the changes.